File: array/python/FractionalKnapsack.py
# Time: O(nlog(n))
# Space: O(1)

class Item:
	def __init__(self, value, weight):
		self.value = value
		self.weight = weight

def fractionalknapsack(W, Items, n):
    def compare(x):
        return x.value / x.weight
    Items.sort(key = compare, reverse = True)
    profit = 0
    for i in range(n):
        if (W >= Items[i].weight):
            W -= Items[i].weight
            profit += Items[i].value
        else:
            profit += (W / Items[i].weight) * Items[i].value
            break
    return profit

if __name__=='__main__':
	test_cases = [[3, 50, [60, 100, 120], [10, 20, 30]],
				  [2, 50, [60, 100], [10, 20]]]
	for n, allowed_weight, values, weight in test_cases:
		Items = []
		for i in range(n):
			Items.append(Item(values[i], weight[i]))
		print(fractionalknapsack(allowed_weight, Items, n))

File: array/java/FractionalKnapsack.java
import java.util.Arrays;

public class FractionalKnapsack {
    public float fractionalknapsack(int[][] Items, float W) {
        float profit = 0;

        Arrays.sort(Items, (a,b)-> (b[1] / b[0]) - (a[1] / a[0]));
        
        for(int i = 0; i < Items.length; i++) {
            if(W >= Items[i][0]) {
                W -= Items[i][0];
                profit += Items[i][1];
            } else {
                profit += (W / Items[i][0]) * Items[i][1];
                break;
            }
        }
        return profit;
    }
    public static void main(String[] args){
        FractionalKnapsack fs = new FractionalKnapsack();
        int Items[][] = new int[][]{{10, 60}, {20, 100}, {30, 120}};
        int W = 50;
        System.out.println(fs.fractionalknapsack(Items, W));
    }
}
